cmd/outline-ss-server/metrics.go

```diff
@@ -49,49 +58,164 @@ type outlineMetrics struct {
+	ipinfo.IPInfoMap
+	mu            sync.Mutex // Protects the activeClients map.
+	activeClients map[IPKey]*activeClient
+
+	tunnelTimePerKey      *prometheus.Desc
+	tunnelTimePerLocation *prometheus.Desc
+}
+
+func (c *tunnelTimeCollector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- c.tunnelTimePerKey
+	ch <- c.tunnelTimePerLocation
+}
+
+func (c *tunnelTimeCollector) Collect(ch chan<- prometheus.Metric) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	tNow := now()
+	for ipKey, client := range c.activeClients {
+		var connDuration = client.connDuration
+		if client.connCount > 0 {
+			connDuration += tNow.Sub(client.startTime)
+		}
+		logger.Debugf("Collecting TunnelTime for key `%v`, duration: %v", ipKey.accessKey, connDuration)
+		ch <- prometheus.MustNewConstMetric(c.tunnelTimePerKey, prometheus.CounterValue, connDuration.Seconds(), ipKey.accessKey)
+		ch <- prometheus.MustNewConstMetric(c.tunnelTimePerLocation, prometheus.CounterValue, connDuration.Seconds(), client.info.CountryCode.String(), asnLabel(client.info.ASN))
+		if client.connCount == 0 {
+			delete(c.activeClients, ipKey)
+			continue
+		}
+		// Reset the timing components now that TunnelTime has been reported.
+		client.startTime = tNow
+		client.connDuration = 0
+	}
+}
+
+// Registers a new active connection for a client [net.Addr] and access key.
+func (c *tunnelTimeCollector) startConnection(ipKey IPKey) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	client, exists := c.activeClients[ipKey]
+	if !exists {
+		clientInfo, _ := ipinfo.GetIPInfoFromIP(c.IPInfoMap, net.IP(ipKey.ip.AsSlice()))
+		client = &activeClient{info: clientInfo}
+	}
+	if client.connCount == 0 {
+		// This client is new or was recently stopped (before it was measured and
+		// deleted by the metrics reporter). Reset the start time so we are
+		// accurately capturing this new session.
+		client.startTime = now()
+	}
+	client.connCount++
+	c.activeClients[ipKey] = client
+}
+
+// Removes an active connection for a client [net.Addr] and access key.
+func (c *tunnelTimeCollector) stopConnection(ipKey IPKey) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	client, exists := c.activeClients[ipKey]
+	if !exists {
+		logger.Warningf("Failed to find active client")
+		return
+	}
+	client.connCount--
+	if client.connCount == 0 {
+		client.connDuration = now().Sub(client.startTime)
+	}
+}
+
+func newTunnelTimeTracker(ip2info ipinfo.IPInfoMap, registerer prometheus.Registerer) *tunnelTimeCollector {
+	c := &tunnelTimeCollector{
+		IPInfoMap:     ip2info,
+		activeClients: make(map[IPKey]*activeClient),
+
+		tunnelTimePerKey: prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, "", "tunnel_time_seconds"),
+			"Time at least 1 connection was open for a (IP, access key) pair, per key.",
+			[]string{"access_key"}, nil,
+		),
+		tunnelTimePerLocation: prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, "", "tunnel_time_seconds_per_location"),
+			"Time at least 1 connection was open for a (IP, access key) pair, per location.",
```

Comment:
I think we can clarify the description. "time at least 1 connection was open for a (IP, access key) pair" is a bit confusing, not very user-friendly. I think users would read that and not quite know what it is.

Perhaps something like "Tunnel time per access key" and "Tunnel time per location and ASN".
Or "active device (based on IP + access key) seconds per key"
Or "Session time per access key".
 
Tunnel time is probably best if we use the name consistently across the product.